// galvanize-game-mechanics/src/castles/kingdom.js
const _ = require('lodash')
const generate = require('project-name-generator')
const uuid = require('uuid/v4')
const util = require('./util')
const Castle = require('./castle')

class Kingdom {
  constructor (name=generate().spaced) {
    this.id = uuid()
    this._castles = []
    this.name = _.startCase(name)
  }

  get castles () {
    return this._castles
  }

  set castles (val) {
    return this._castles
  }

  createCastle (name) {
    if(this.castles.length < 3) {
      const castle = new Castle(name)
      this._castles.push(castle)
    }

    return this
  }

  _modifyCastle (id, { attack, build }) {
    if (!attack && !build) throw new Error('Must specify whether to attack or build castle')
    if (attack && build) throw new Error('Can only attack or build; not both')

    const castle = _.find(this.castles, { id })
    if (!castle) return false

    if (attack && !build) castle.health -= util.attack()
    if (!attack && build) castle.health += util.build()

    return castle
  }

  attackCastle (id) {
    return this._modifyCastle(id, { attack: true })
  }

  buildCastle (id) {
    return this._modifyCastle(id, { build: true })
  }
}

module.exports = Kingdom
